### PR TITLE
Add saturation-aware forced rotation for efficiency optimization

### PR DESCRIPTION
**Pull request by tomquist**
## Summary
Implements saturation-aware forced rotation in the efficiency optimization system. When an active battery cannot follow its power target (detected via saturation), it is automatically swapped with a healthy deprioritized battery, improving system responsiveness to constraints like external power limits or SOC boundaries.

### Core Feature (CT002)
- **Saturation detection and decay**: Added `_update_saturation()` logic to detect when a battery cannot follow its target (saturation) and decay the saturation metric over time when the target is below the minimum threshold
- **Forced swap mechanism**: Implemented `_maybe_force_swap_saturated()` to automatically rotate a saturated active battery with a healthy deprioritized one when saturation exceeds a configurable threshold
- **Configuration parameters**:
  - `efficiency_saturation_threshold` (0.0–1.0): Saturation level that triggers forced rotation (0.0 disables the feature)
  - `saturation_decay_factor` (0.0–1.0): Multiplicative decay rate for saturation when target is below minimum
- **Cache invalidation**: Forced swaps invalidate the efficiency cache to ensure immediate effect on subsequent consumer calls

### Configuration & Integration
- Added config file parameters (`EFFICIENCY_SATURATION_THRESHOLD`, `SATURATION_DECAY_FACTOR`) with sensible defaults
- Integrated parameters into `main.py` and simulator runner
- Added `time_scale` parameter to battery simulator for accelerated testing

### Testing
- **Unit tests** (`test_ct002_active_control.py`): 10 new tests covering saturation detection, forced swap logic, decay, edge cases (all saturated, disabled feature), and multi-consumer scenarios
- **E2E tests** (`test_efficiency_e2e.py`): 2 new async tests validating saturated battery rotation and recovery after constraint lift
- **Smoke tests** (new files):
  - `smoke_efficiency_saturation.py`: 8 comprehensive scenarios (basic swap, recovery, no ping-pong, 3 batteries, load changes, both constrained, feature disabled, charging direction)
  - `smoke_transitions.py`: 7 post-stabilization transition scenarios (load steps, solar ramps, SOC boundaries, multi-phase shifts)

## Implementation Details
- Saturation is calculated as the ratio of unmet demand to target when direction reversal occurs (e.g., target discharge but actual charge)
- Forced swaps only occur if there is a healthy (non-saturated) replacement available; if all consumers are saturated, no swap happens
- Rotation timer is reset on forced swap to prevent immediate re-rotation
- Saturation entries decay multiplicatively and are removed when they fall below 0.001 to prevent stale state
- Feature is opt-in via threshold parameter; setting to 0.0 disables all forced swaps

https://claude.ai/code/session_01J8ittzuocCdaGZWm9hz88Q

## Summary by CodeRabbit

* **New Features**
  * Battery saturation detection now forces immediate swap of saturated batteries during efficiency optimization (default threshold 0.4).
  * Configurable saturation decay factor to control how quickly swapped batteries become eligible again.

* **Simulator**
  * Added time-scaling option to accelerate or slow simulated time.

* **Documentation**
  * README, example config, and changelog updated with threshold and decay details.

* **Tests**
  * Extensive new smoke and unit tests covering saturation-driven swaps, recovery, timing, and edge cases.

### CODE REVIEW

File: src/b2500_meter/ct002/ct002.py [hunk 1]
```diff
@@ -385,6 +456,19 @@ def _compute_efficiency_deprioritized(self, reports, sample_id):
...
+        # Must happen BEFORE the forced swap check.
+        for cid in self._efficiency_deprioritized - deprioritized:
+            self._saturation_by_consumer.pop(cid, None)
+
+        if self._maybe_force_swap_saturated(self._efficiency_priority, slots, now):
+            # Recompute after swap.  Cache naturally invalidates because
+            # cache_key (built above with the pre-swap priority tuple) won't
+            # match the next caller's key (built with the post-swap tuple).
+            deprioritized = set(self._efficiency_priority[slots:])
+            result = {cid: 0.0 for cid in deprioritized}
```
[the remark this answers is not in the document]
Comment (tomquist) [reply src/b2500_meter/ct002/ct002.py:470]: Fixed in f2c7256. Both issues addressed:

1. **`cache_key` rebuild** — `cache_key` is now rebuilt from the post-swap `self._efficiency_priority` after a forced swap, so subsequent consumers in the same sample can hit the cache with the correct key.
2. **Saturation clearing for swap-promoted consumers** — was already added in 8dd6ce7 (lines 471-474): after the swap, saturation is cleared for consumers that moved from deprioritized to active (`self._efficiency_deprioritized - deprioritized`).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/smoke_efficiency_saturation.py [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+            ct_mac=ct_mac,
+            active_control=True,
+            fair_distribution=True,
+            smooth_target_alpha=0.9,
+            deadband=5,
+            min_efficient_power=min_efficient_power,
+            efficiency_rotation_interval=efficiency_rotation_interval / time_scale,
+            efficiency_saturation_threshold=efficiency_saturation_threshold,
+            saturation_decay_factor=saturation_decay_factor,
+            consumer_ttl=120 / time_scale,
```
[the remark this answers is not in the document]
Comment (tomquist) [reply tests/smoke_efficiency_saturation.py:118]: Partially addressed in f2c7256:

- **`time_scale > 0` guard** — added, raises `ValueError` on non-positive values.
- **`scaled_rotation` pre-computed** — the division is done once into a named variable for clarity.
- **Floor check omitted** — several smoke scenarios intentionally pass `efficiency_rotation_interval=15` or `30` with `time_scale=10`, producing scaled values of 1.5s–3s. These deliberately rely on CT002's `max(10, ...)` clamp to get fast rotation for testing. Adding a floor check here would break those scenarios. CT002 handles its own clamping safely.

---
_Generated by [Claude Code](https://claude.ai/code)_
